src/main/java/net/minecraftforge/datafix/ForgeSchema.java

```diff
@@ -0,0 +1,340 @@
+        {
+            parent.registerTypes(schema, entityTypes, blockEntityTypes);
+        }
+    }
+
+    @Override
+    public Map<String, Supplier<TypeTemplate>> registerEntities(final Schema schema)
+    {
+        //This is a bit wonky, in vanilla it suffices to register the types to one single version
+        //And all newer versions will have this type.
+        //Due to the way we set up the schemas with events later, that is not possible anymore.
+        //Modders will need to register their types on all versions that they should exist in,
+        //not just in the oldest version.
+
+        if (wrapped != null)
+        {
+            return wrapped.registerEntities(schema);
+        }
+        if (parent != null)
+        {
+            return parent.registerEntities(schema);
+        }
+
+        return Maps.newHashMap();
+    }
+
+    @Override
+    public Map<String, Supplier<TypeTemplate>> registerBlockEntities(final Schema schema)
+    {
+        //This is a bit wonky, in vanilla it suffices to register the types to one single version
+        //And all newer versions will have this type.
+        //Due to the way we set up the schemas with events later, that is not possible anymore.
+        //Modders will need to register their types on all versions that they should exist in,
+        //not just in the oldest version.
+
+        if (wrapped != null)
+        {
+            return wrapped.registerBlockEntities(schema);
+        }
+        if (parent != null)
+        {
+            return parent.registerBlockEntities(schema);
+        }
+
+        return Maps.newHashMap();
+    }
+
+    /**
+     * Registers a new type to the Schema with its type template.
+     *
+     * @param recursive Indicates if the type is potentially recursive.
+     * @param type The type in question.
+     * @param template Its template.
+     */
+    @Override
+    public void registerType(final boolean recursive, final DSL.TypeReference type, final Supplier<TypeTemplate> template)
+    {
+        if (TYPE_TEMPLATES == null)
+            TYPE_TEMPLATES = Maps.newHashMap();
+
+        TYPE_TEMPLATES.put(type.typeName(), template);
+        if (recursive && !RECURSIVE_TYPES.containsKey(type.typeName()))
+        {
+            RECURSIVE_TYPES.put(type.typeName(), RECURSIVE_TYPES.size());
+        }
+    }
+
+    /**
+     * Resets the schema so that the type map can be properly rebuild.
+     */
+    public void resetSchema()
+    {
+        this.TYPE_TEMPLATES.clear();
+        this.TYPES.clear();
+        this.RECURSIVE_TYPES.clear();
+        this.MODDED_ENTITY_TYPES.clear();
+        this.MODDED_BLOCK_ENTITY_TYPES.clear();
+    }
+
+    /**
+     * Rebuilds the schema's type map.
+     * This also injects custom types for entities and block entities.
+     *
+     * @param modEntityTypes The mod entity types to inject.
+     * @param modBlockEntityTypes The block entity types to inject.
+     */
+    public void rebuildSchema(
```

Comment:
I feel like this should be package private, especially then if we add more support for other types vanilla doesn't, we can easily change the signature. (Technically it doesn't matter given the class itself is package private but still.)